src/game.rs

```diff
@@ -21,90 +21,237 @@ pub enum ChessError {
     #[default]
     Ongoing,
     Stalemate,
     Check,
     Checkmate,
+    Replay,
+}
+
+impl GameStatus {
+    pub fn update(&mut self, status: GameStatus) {
+        if *self != status {
+            println!("GameStatus changing from {:?} to {:?}", *self, status)
+        }
+        *self = status;
+    }
 }
 
 #[derive(Clone, Default)]
+pub struct History {
+    history: Vec<(BoardState, Move)>,
+    current_turn: usize,
+}
+
+impl History {
+    pub fn with_state(state: BoardState) -> Self {
+        Self {
+            history: vec![(state, Move::default())],
+            ..Default::default()
+        }
+    }
+
+    fn add_info(&mut self, next_state: BoardState, mv: Move) {
+        self.history.push((next_state, mv));
+        self.current_turn += 1
+    }
+
+    fn get_current_state(&self) -> &BoardState {
+        &self.history[self.current_turn - 1].0
+    }
+
+    fn get_info_for_turn(&self, turn: usize) -> &(BoardState, Move) {
+        &self.history[turn]
+    }
+
+    fn resume(&mut self) {
+        self.current_turn = self.history.len()
+    }
+
+    fn previous_state(&mut self) {
+        if self.current_turn > 1 {
+            self.current_turn -= 1
+        }
+    }
+
+    fn next_state(&mut self) {
+        if self.current_turn < self.history.len() {
+            self.current_turn += 1
+        }
+    }
+
+    fn initial_state(&mut self) {
+        self.current_turn = 1
+    }
+
+    fn is_ongoing(&mut self) -> bool {
+        self.current_turn == self.history.len()
+    }
+}
+
+#[derive(Clone)]
 pub struct Game {
-    state: BoardState,
     valid_moves: HashSet<Move>,
     status: GameStatus,
+    history: History,
+}
+
+impl Default for Game {
+    fn default() -> Self {
+        Game::with_history(History::with_state(BoardState::default()))
+    }
 }
 
 impl Game {
     pub fn new() -> Self {
-        let mut game = Self::default();
+        Game::default()
+    }
+
+    pub fn with_history(history: History) -> Self {
+        let mut game = Self {
+            history,
+            ..Default::default()
+        };
         game.add_moves();
         game
     }
 
+    pub fn with_state(state: BoardState) -> Self {
+        Game::with_history(History::with_state(state))
+    }
+
     pub fn get_piece(&self, position: &Position) -> Option<Piece> {
-        self.state.get_piece(position)
+        self.get_current_state().get_piece(position)
     }
 
-    pub fn get_board_state(&self) -> &BoardState {
-        &self.state
+    fn get_current_state(&self) -> &BoardState {
+        self.history.get_current_state()
+    }
+
+    fn get_current_player(&self) -> Player {
+        self.get_current_state().player
+    }
+
+    fn is_piece_some(&self, at: &Position) -> ChessResult {
+        self.get_current_state().is_piece_some(at)
+    }
+
+    pub fn has_piece(&self, position: &Position) -> bool {
+        self.get_current_state().has_piece(position)
+    }
+
+    fn piece_can_snipe(&self, at: &Position) -> bool {
```

Comment:
Is there any magic we could do with `?`? If this is going to be a general function, maybe we shouldn't be `unwrap`ing.